src/mozanalysis/experiments.py

```diff
@@ -1,165 +1,268 @@
+p5.__name__ = "p5"
+
+p25 = partial(np.percentile, q=25)
+p25.__name__ = "p25"
+
+p50 = partial(np.percentile, q=50)
+p50.__name__ = "p50"
+
+p75 = partial(np.percentile, q=75)
+p75.__name__ = "p75"
+
+p95 = partial(np.percentile, q=95)
+p95.__name__ = "p95"
 
 
 class ExperimentAnalysis(object):
-    def __init__(self, dataset):
-        self.sc = dataset._sc
-        self._dataset = dataset
-        self._metrics = None
-        self._date_aggregate_by = "submission_date_s3"
-        self._aggregate_by = "client_id"
-        self._split_by = "experiment_branch"
-
-    def metrics(self, *metrics):
-        self._metrics = metrics
-        return self
-
-    def date_aggregate_by(self, column):
-        self._date_aggregate_by = column
-        return self
-
-    def aggregate_by(self, column):
-        self._aggregate_by = column
-        return self
-
-    def split_by(self, column):
-        self._split_by = column
-        return self
-
-    def run(self):
-        """
-        Run the full analysis on the provided dataset.
-
-        This will perform the per-client per-day aggregation, as well as the
-        analysis aggregations, across all metrics.
-
-        """
-        df = self.aggregate_per_client_daily(self._dataset)
-        return self.analyze(df)
-
-    def get_split_by_values(self, dataset):
-        """
-        Returns the dictinct column values of the `split_by` column.
-
-        """
-        return [
-            r[self._split_by]
-            for r in dataset.select(self._split_by).distinct().collect()
-        ]
+    def __init__(self, spark_session, experiment_id, stats=None):
+        self.spark_session = spark_session
+        self.experiment_id = experiment_id
+        self._df = None
 
-    def aggregate_per_client_daily(self, dataset):
-        """
-        Returns a pyspark dataframe of the data aggregated by the column
-        defined in the `date_aggregate_by` call, which is
-        "submission_date_s3" by default.
-
-        """
-        cols = [
-            F.col(self._aggregate_by),
-            F.col(self._split_by),
-            F.col(self._date_aggregate_by),
-        ]
-        aggs = []
-        for m in self._metrics:
-            cols.extend(m.daily_columns)
-            aggs.extend(m.daily_aggregations)
+        if stats:
+            if isinstance(stats, (list, tuple)):
+                self.stats = stats
+            else:
+                self.stats = [stats]
+        else:
+            self.stats = [np.mean, p5, p25, p50, p75, p95]
 
-        cols = dedupe_columns(cols)
-        aggs = dedupe_columns(aggs)
+    @property
+    def df(self):
+        if self._df:
+            return self._df
+
+        return self.get_data()
+
+    @df.setter
+    def df(self, value):
+        self._df = value
+
+    def get_experiment_data(self):
+        df = self.spark_session.read.parquet("s3://telemetry-parquet/experiments/v1")
+        df = df.filter(F.col("experiment_id") == self.experiment_id)
```

Comment:
It should be possible for the caller to specify this filtered table, perhaps as an argument to the constructor.

Using `df` for the name of this object is also a little confusing since class instances have a `df` property and this isn't that

Reply:
What do you mean by "this filtered table"? The s3 parquet path?

Reply:
I would like to be able to, at least optionally, pass in a Spark DataFrame which contains the rows of a table matching the `experiments` schema that are relevant to the experiment, instead of letting the class build that SDF itself from its own knowledge.

Reply:
Okay. That's what I had intended for the property setter on the `_df` variable. If you look at the tests it uses this to assign a dataframe for testing. The same could be done in a notebook, e.g., to assign the dataframe and still call `.engagement()` on it. Does that satisfy that use case?

Reply:
That would work except that `ExperimentAnalysis.df` holds the result of `get_data`, which is a different data frame which has already been aggregated to the daily level, so it doesn't satisfy this use case.

Reply:
Can you elaborate on what the API would look like? Is this a requirement for this PR? I'm having trouble understanding exactly what you're expecting. The `get_data` call calls `get_experiment_data` and then `aggregate_per_day`. If you're saying that you want to provide a SDF that gets aggregated per day, that could be done via something like the following:

```
ea = ExperimentAnalysis(spark, "experiment-slug")
agg_df = ea.aggregate_per_day(my_sdf)
ea.df = agg_df
ea.retention()
...
```